File: crud.go
package main

import (
	"database/sql"
	"fmt"
	"log"

	_ "github.com/mattn/go-sqlite3"
)

func create() {
	db, err := sql.Open("sqlite3", "data.db")
	if err != nil {
		log.Fatal(err)
	}
	defer db.Close()

	createTable := `
		CREATE TABLE IF NOT EXISTS Users (
			ID      INTEGER PRIMARY KEY,
			NAME    TEXT NOT NULL,
			AGE     INTEGER NOT NULL,
			GENDER  TEXT,
			SALARY  INTEGER
		);
	`
	_, err = db.Exec(createTable)
	if err != nil {
		log.Fatal(err)
	}

	fmt.Println("Table created successfully")
}

func insert() {
	db, err := sql.Open("sqlite3", "data.db")
	if err != nil {
		log.Fatal(err)
	}
	defer db.Close()

	stmt, err := db.Prepare("INSERT INTO Users(ID, NAME, AGE, GENDER, SALARY) VALUES (?, ?, ?, ?, ?)")
	if err != nil {
		log.Fatal(err)
	}

	var id, age, salary int
	var name, gender string

	fmt.Print("Enter ID: ")
	fmt.Scan(&id)
	fmt.Print("Enter Name: ")
	fmt.Scan(&name)
	fmt.Print("Enter Age: ")
	fmt.Scan(&age)
	fmt.Print("Enter Gender: ")
	fmt.Scan(&gender)
	fmt.Print("Enter Salary: ")
	fmt.Scan(&salary)

	_, err = stmt.Exec(id, name, age, gender, salary)
	if err != nil {
		log.Fatal(err)
	}

	fmt.Println("Data Inserted Successfully")
}

func readOne() {
	db, err := sql.Open("sqlite3", "data.db")
	if err != nil {
		log.Fatal(err)
	}
	defer db.Close()

	var id int
	fmt.Print("Enter Your ID: ")
	fmt.Scan(&id)

	row := db.QueryRow("SELECT * FROM Users WHERE ID = ?", id)

	var name, gender string
	var age, salary int

	err = row.Scan(&id, &name, &age, &gender, &salary)
	if err != nil {
		log.Fatal(err)
	}

	fmt.Printf("Name is: %s\n", name)
	fmt.Printf("Age is: %d\n", age)
	fmt.Printf("Salary is: %d\n", salary)
}

func readAll() {
	db, err := sql.Open("sqlite3", "data.db")
	if err != nil {
		log.Fatal(err)
	}
	defer db.Close()

	rows, err := db.Query("SELECT * FROM Users")
	if err != nil {
		log.Fatal(err)
	}
	defer rows.Close()

	fmt.Println("\n<=== Available Records ===>")
	for rows.Next() {
		var id, age, salary int
		var name, gender string

		err = rows.Scan(&id, &name, &age, &gender, &salary)
		if err != nil {
			log.Fatal(err)
		}

		fmt.Printf("Name is: %s\n", name)
		fmt.Printf("Age is: %d\n", age)
		fmt.Printf("Salary is: %d\n", salary)
	}
}

func update() {
	db, err := sql.Open("sqlite3", "data.db")
	if err != nil {
		log.Fatal(err)
	}
	defer db.Close()

	stmt, err := db.Prepare("UPDATE Users SET NAME = ?, AGE = ?, GENDER = ?, SALARY = ? WHERE ID = ?")
	if err != nil {
		log.Fatal(err)
	}

	var id, age, salary int
	var name, gender string

	fmt.Print("Enter ID: ")
	fmt.Scan(&id)
	fmt.Print("Enter Name: ")
	fmt.Scan(&name)
	fmt.Print("Enter Age: ")
	fmt.Scan(&age)
	fmt.Print("Enter Gender: ")
	fmt.Scan(&gender)
	fmt.Print("Enter Salary: ")
	fmt.Scan(&salary)

	_, err = stmt.Exec(name, age, gender, salary, id)
	if err != nil {
		log.Fatal(err)
	}

	fmt.Println("Records Updated")
}

func delete() {
	db, err := sql.Open("sqlite3", "data.db")
	if err != nil {
		log.Fatal(err)
	}
	defer db.Close()

	stmt, err := db.Prepare("DELETE FROM Users WHERE ID = ?")
	if err != nil {
		log.Fatal(err)
	}

	var id int
	fmt.Print("Enter ID: ")
	fmt.Scan(&id)

	_, err = stmt.Exec(id)
	if err != nil {
		log.Fatal(err)
	}

	fmt.Println("One record Deleted")
}

func main() {
	create()

	for {
		fmt.Println("1). Insert Records")
		fmt.Println("2). Read Records")
		fmt.Println("3). Update Records")
		fmt.Println("4). Delete Records")
		fmt.Println("5). Exit")

		var ch int
		fmt.Print("Enter Your Choice: ")
		fmt.Scan(&ch)

		switch ch {
		case 1:
			insert()
		case 2:
			fmt.Println("1). Read Single Record")
			fmt.Println("2). Read All Records")

			var choice int
			fmt.Print("Enter Your Choice: ")
			fmt.Scan(&choice)

			switch choice {
			case 1:
				readOne()
			case 2:
				readAll()
			default:
				fmt.Println("Wrong Choice Entered")
			}
		case 3:
			update()
		case 4:
			delete()
		case 5:
			return
		default:
			fmt.Println("Enter Correct Choice")
		}
	}
}

File: crud.js
const sqlite3 = require('sqlite3').verbose();
const readline = require('readline');

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

const db = new sqlite3.Database('data.db');

function create() {
  db.serialize(() => {
    db.run(`
      CREATE TABLE IF NOT EXISTS Users (
        ID INTEGER PRIMARY KEY NOT NULL,
        NAME TEXT NOT NULL,
        AGE INTEGER NOT NULL,
        GENDER TEXT,
        SALARY INTEGER
      )
    `, (err) => {
      if (err) {
        console.error('Table creation error:', err);
      } else {
        console.log('Table created successfully');
      }
    });
  });
}

function insert() {
  rl.question('Enter ID: ', (idd) => {
    rl.question('Enter Name: ', (name) => {
      rl.question('Enter Age: ', (age) => {
        rl.question('Enter your Gender: ', (gender) => {
          rl.question('Enter your Salary: ', (salary) => {
            const query = 'INSERT INTO Users (ID, NAME, AGE, GENDER, SALARY) VALUES (?, ?, ?, ?, ?)';
            const data = [idd, name, age, gender, salary];
            db.run(query, data, (err) => {
              if (err) {
                console.error('Data insertion error:', err);
              } else {
                console.log('Data inserted successfully');
              }
              rl.close();
            });
          });
        });
      });
    });
  });
}

function readOne() {
  rl.question('Enter Your ID: ', (ids) => {
    const query = 'SELECT * FROM Users WHERE ID = ?';
    db.get(query, [ids], (err, row) => {
      if (err) {
        console.error('Read one record error:', err);
      } else if (row) {
        console.log(`Name is: ${row.NAME}`);
        console.log(`Age is: ${row.AGE}`);
        console.log(`Salary is: ${row.SALARY}`);
      } else {
        console.log('Record does not exist');
      }
      rl.close();
    });
  });
}

function readAll() {
  const query = 'SELECT * FROM Users';
  db.all(query, (err, rows) => {
    if (err) {
      console.error('Read all records error:', err);
    } else if (rows.length > 0) {
      console.log('\n<=== Available Records ===>');
      rows.forEach((row) => {
        console.log(`Name is: ${row.NAME}`);
        console.log(`Age is: ${row.AGE}`);
        console.log(`Salary is: ${row.SALARY}\n`);
      });
    } else {
      console.log('No records found');
    }
    rl.close();
  });
}

function update() {
  rl.question('Enter ID: ', (idd) => {
    rl.question('Enter Name: ', (name) => {
      rl.question('Enter Age: ', (age) => {
        rl.question('Enter Gender: ', (gender) => {
          rl.question('Enter Salary: ', (salary) => {
            const query = 'UPDATE Users SET NAME = ?, AGE = ?, GENDER = ?, SALARY = ? WHERE ID = ?';
            const data = [name, age, gender, salary, idd];
            db.run(query, data, (err) => {
              if (err) {
                console.error('Record update error:', err);
              } else {
                console.log('Record updated successfully');
              }
              rl.close();
            });
          });
        });
      });
    });
  });
}

function remove() {
  rl.question('Enter ID: ', (idd) => {
    const query = 'DELETE FROM Users WHERE ID = ?';
    db.run(query, [idd], (err) => {
      if (err) {
        console.error('Record deletion error:', err);
      } else {
        console.log('Record deleted successfully');
      }
      rl.close();
    });
  });
}

function main() {
  rl.question('1) Insert Records\n2) Read Records\n3) Update Records\n4) Delete Records\n5) Exit\nEnter Your Choice: ', (ch) => {
    switch (ch) {
      case '1':
        insert();
        break;
      case '2':
        rl.question('1) Read Single Record\n2) Read All Records\nEnter Your Choice: ', (choice) => {
          if (choice === '1') {
            readOne();
          } else if (choice === '2') {
            readAll();
          } else {
            console.log('Wrong choice entered');
            rl.close();
          }
        });
        break;
      case '3':
        update();
        break;
      case '4':
        remove();
        break;
      case '5':
        rl.close();
        break;
      default:
        console.log('Enter correct choice');
        rl.close();
        break;
    }
  });
}

create();
main();

process.on('exit', () => {
  db.close();
});
